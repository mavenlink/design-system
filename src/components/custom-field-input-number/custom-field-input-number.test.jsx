import React, { createRef } from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import CustomFieldInputNumber from './custom-field-input-number.jsx';

describe('CustomFieldInputNumber', () => {
  function TestComponent(props = {}) {
    return <CustomFieldInputNumber id="test-input" label="Test label" {...props} />;
  }

  it('has defaults', () => {
    const tree = renderer.create(<TestComponent />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('className API', () => {
    it('prioritizes className prop', () => {
      const { container } = render(<TestComponent className="prioritize-me" />);
      expect(container.firstChild).toHaveClass('prioritize-me');
    });
  });

  describe('disabled API', () => {
    it('can be disabled', () => {
      const { container } = render(<TestComponent disabled />);
      expect(container.firstChild).toHaveClass('disabled');
      expect(screen.getByLabelText('Test label')).toBeDisabled();
    });

    it('can be enabled', () => {
      const { container } = render(<TestComponent />);
      expect(container.firstChild).not.toHaveClass('disabled');
      expect(screen.getByLabelText('Test label')).not.toBeDisabled();
    });
  });

  describe('id API', () => {
    it('sets the id attribute', () => {
      render(<TestComponent id="test-id" />);
      expect(screen.getByLabelText('Test label')).toHaveAttribute('id', 'test-id');
    });
  });

  describe('label API', () => {
    it('sets the label', () => {
      render(<TestComponent label="Another label" />);
      expect(screen.getByLabelText('Another label')).toBeDefined();
    });
  });

  describe('name API', () => {
    it('sets the name attribute', () => {
      render(<TestComponent name="test-name" />);
      expect(screen.getByLabelText('Test label')).toHaveAttribute('name', 'test-name');
    });
  });

  describe('placeholder API', () => {
    it('sets the placeholder attribute', () => {
      const placeholder = 'This is placeholder input';
      render(<TestComponent placeholder={placeholder} />);
      expect(screen.getByLabelText('Test label')).toHaveAttribute('placeholder', placeholder);
    });
  });

  describe('required API', () => {
    it('sets the required attribute', () => {
      render(<TestComponent required />);
      expect(screen.getByLabelText('Test label')).toBeRequired();
    });

    it('unsets the required attribute', () => {
      render(<TestComponent />);
      expect(screen.getByLabelText('Test label')).not.toBeRequired();
    });
  });

  describe('value API', () => {
    it('is valid on a positive integer', () => {
      render(<TestComponent value={1} />);
      expect(screen.getByTestId('custom-field-input')).not.toHaveClass('error');
    });

    it('is valid on zero', () => {
      render(<TestComponent value={0} />);
      expect(screen.getByTestId('custom-field-input')).not.toHaveClass('error');
    });

    it('is valid on a negative integer', () => {
      render(<TestComponent value={-1} />);
      expect(screen.getByTestId('custom-field-input')).not.toHaveClass('error');
    });

    it('is valid on a decimal integer', () => {
      render(<TestComponent value={1.00} />);
      expect(screen.getByTestId('custom-field-input')).not.toHaveClass('error');
    });

    it('is invalid on a decimal number', () => {
      render(<TestComponent value={1.01} />);
      expect(screen.getByTestId('custom-field-input')).toHaveClass('error');
    });
  });

  describe('events API', () => {
    it('accepts an onBlur event', () => {
      const onBlur = jest.fn();
      const { getByLabelText } = render(<TestComponent label="foo" onBlur={onBlur} />);
      fireEvent.blur(getByLabelText('foo'));
      expect(onBlur.mock.calls.length).toEqual(1);
    });
  });

  describe('inputRef API', () => {
    it('sets the ref on the input', () => {
      const inputRef = createRef();
      render(<TestComponent inputRef={inputRef} />);
      expect(screen.getByLabelText('Test label')).toBe(inputRef.current);
    });
  });

  describe('step API', () => {
    it('respects a provided step', () => {
      const { getByLabelText } = render(<TestComponent label="foo" step={12} />);
      expect(getByLabelText('foo')).toHaveAttribute('step', '12');
    });
  });
});
