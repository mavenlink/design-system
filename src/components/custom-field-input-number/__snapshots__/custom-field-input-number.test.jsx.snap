// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomFieldInputNumber has defaults 1`] = `
<body>
  <div>
    <div
      role="presentation"
    >
      <label
        class="label"
        for="test-input"
      >
        Test label
      </label>
      <div
        class="control-container"
      >
        <input
          aria-describedby="test-inputHint"
          class="custom-field-input-text"
          id="test-input"
          name="field-id"
          step="1"
          type="number"
          value=""
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`CustomFieldInputNumber has defaults 2`] = `
Object {
  "dirty": true,
  "focus": [Function],
  "id": "test-input",
  "name": "field-id",
  "validity": ValidityState {},
  "value": "NaN",
}
`;
